sway-core/src/asm_generation/asm_builder.rs

```diff
@@ -492,33 +491,44 @@ impl<'ir> AsmBuilder<'ir> {
     fn compile_conditional_branch(
         &mut self,
         cond_value: &Value,
-        from_block: &Block,
-        true_block: &Block,
-        false_block: &Block,
-    ) {
-        self.compile_branch_to_phi_value(from_block, true_block);
-        self.compile_branch_to_phi_value(from_block, false_block);
+        true_block: &BranchToWithArgs,
+        false_block: &BranchToWithArgs,
+    ) -> CompileResult<()> {
+        if true_block.block == false_block.block && true_block.block.num_args(self.context) > 0 {
+            return err(
+                Vec::new(),
+                vec![CompileError::Internal(
+                    "Cannot compile CBR with both branches going to same dest block",
+                    self.md_mgr
+                        .val_to_span(self.context, *cond_value)
+                        .unwrap_or_else(Self::empty_span),
+                )],
+            );
+        }
+        self.compile_branch_to_phi_value(true_block);
+        self.compile_branch_to_phi_value(false_block);
 
         let cond_reg = self.value_to_register(cond_value);
 
-        let true_label = self.block_to_label(true_block);
+        let true_label = self.block_to_label(&true_block.block);
         self.cur_bytecode
             .push(Op::jump_if_not_zero(cond_reg, true_label));
 
-        let false_label = self.block_to_label(false_block);
+        let false_label = self.block_to_label(&false_block.block);
         self.cur_bytecode.push(Op::jump_to_label(false_label));
+        ok((), vec![], vec![])
     }
 
-    fn compile_branch_to_phi_value(&mut self, from_block: &Block, to_block: &Block) {
-        if let Some(local_val) = to_block.get_phi_val_coming_from(self.context, from_block) {
-            // We only need a MOVE here if get_phi_val_coming_from() is actually assigned to a
-            // register
-            if let Some(local_reg) = self.opt_value_to_register(&local_val) {
-                let phi_reg = self.value_to_register(&to_block.get_phi(self.context));
+    fn compile_branch_to_phi_value(&mut self, to_block: &BranchToWithArgs) {
```

Comment:
Does it still make sense to call this function `compile_branch_to_phi_value`? Same goes for `phi_reg` below.

Reply:
I considered renaming it, but couldn't think of a better name, and conceptually, it's still a PHI. So I let it be.